daemons/ipa-sam/ipa_sam.c

```diff
@@ -992,6 +994,49 @@ static bool ipasam_uid_to_sid(struct pdb_methods *methods, uid_t uid,
 	return ret;
 }
 
+typedef NTSTATUS (*process_entry_fn)(struct ipasam_private *priv, LDAPMessage *entry, TALLOC_CTX *ctx, void *state);
+static NTSTATUS ipasam_search_entries(struct pdb_methods *methods,
+				      struct ipasam_private *priv,
+				      TALLOC_CTX *ctx,
+				      char *filter,
+				      const char **attrs_list,
+				      process_entry_fn process_entry,
+				      void *state)
+{
+	LDAPMessage *result = NULL;
+	LDAPMessage *entry = NULL;
+	int rc;
+	NTSTATUS res;
+
+	if ((priv == NULL) || (ctx == NULL) || process_entry == NULL) {
+		return NT_STATUS_NO_MEMORY;
+	}
+
+	rc = smbldap_search_suffix(priv->ldap_state, filter, attrs_list, &result);
+	if (rc != LDAP_SUCCESS) {
+		return NT_STATUS_UNSUCCESSFUL;
+	}
+
+	smbldap_talloc_autofree_ldapmsg(ctx, result);
+
+	if (ldap_count_entries(priv2ld(priv), result) == 0) {
+		return NT_STATUS_UNSUCCESSFUL;
+	}
+
+	for (entry = ldap_first_entry(priv2ld(priv), result);
+		entry != NULL;
+		entry = ldap_next_entry(priv2ld(priv), entry)) {
+
+		res = process_entry(priv, entry, ctx, state);
+		if (!NT_STATUS_IS_OK(res)) {
+			return res;
+		}
+	}
+
+	return NT_STATUS_OK;
```

Comment:
result is not freed in either the success or failure case. I think this needs ldap_msgfree(result);

Reply:
We use `smbldap_talloc_autofree_ldapmsg()` which will free the results upon `talloc_free()` on the context. This is standard Samba flow. This happens right after the call to `ipasam_search_entries()`:

```
        result = ipasam_search_entries(methods, ipasam_state, tmp_ctx,
                                       filter, attrs_list,
                                       _ipasam_collect_map_entry, map);
        talloc_free(tmp_ctx);

```